### ISSUE TITLE
Reliability and Performance Dashboard

### ISSUE THREAD

**Issue by Titan-Node** [livepeer/grants#82]
**The Problem:**
A number of potential broadcasters have inquired about the reliability, performance and cost associated with broadcasting through Livepeer. Two individuals in particular were looking to switch from their current web2 solution to a more decentralized and scalable solution. Livepeer fits this description very well; unfortunately, they were unable to find information about these important metrics, most notably reliability.
Upon further discussion a couple key points were discovered by the inquiring broadcasters that were not clear:

1. You can choose specific Orchestrators
They were unaware that customized broadcasting metrics can be configured such as selecting specific Orchestrators that fit their needs. We also spoke about setting up custom pricing and requirements which helped them understand the selection process.

2. All orchestrators are different
Some orchestrators optimize for performance, some optimize for cost, and some don’t optimize in general. After finding out they can choose specific Orchestrators their next question naturally was “Where can I find these metrics” to which we found this question hard to answer.

3. Livepeer is more of a marketplace than a service
These individuals wanted to compare their current web2 service such as AWS with Livepeer, but really they should be comparing AWS with a number of high performing Orchestrators that can each individually compete.
In the end, there seemed to be a need for a dashboard that would display thorough performance metrics that can easily be benchmarked against current web2 solutions.
 

**The Solution:**
A community driven Reliability and Performance Dashboard for helping Broadcasters with their selection process.
Explorer metrics would include:
- Live feed from each Orchestrator (that opt in)
- Trailing 30d Segment Uptime
- Trailing 30d Cost  (USD)
- A comparison to web2 competitors
- Round Trip metrics (distance, up, down, delay)

**Potential features to add later:**
Tool to provide a selection algorithm (frontend)
Search/filter functions (frontend)

As for the 24/7 feed from each Orchestrator this can be accomplished in a number of ways.
These were the options discussed:

**Method 1: (voted for)**
Go-Livepeer could be modified to accept a 24/7 test stream free of cost. This would allow a test Broadcaster to stream into each node and provide a feed free of charge.

**Method 2: (voted against)**
Each Orchestrator could opt-in by running a node on testnet using tentnet ETH.

**Method 3: (voted against)**
The grant pays for the stream cost to each Orchestrator.
 

**The Team:**
The following community members have committed to working on or advising on this project:
- Titan Node
- Papa Bear
- Authority Null
- Mike Zupper
- Vires In Numeris
- Syrinx Node
- Ftkuhnsman
- Strong...like bull
- 0xB79
- EliteEncoder
- Marco
- Pon
- Monitor_ap
- Nightnode

**Deliverables:**
**Phase 1:**
**Title:** Orchestrator side
**Developer Hourly Wage:** $58/hr USD
**Estimated hours:** 40 hours
**Budget:** $2,320.00 USD
**Developers:** 0xB79 and Elite Encoder
**Tasks rendered:**
- Create an array of `-pricePerPixel` options for a list of Broadcaster addresses that can be individually set.

**Phase 2:**
**Title:** Server side / Broadcaster Side
**Developer Hourly Wage:** $58/hr USD* Up for discussion 
**Estimated hours:** 100 hours
**Budget:** $5,800.00 USD
**Developers:** TBA
**Tasks rendered:**
- On rented VPS, for each Orchestrator
   - Customized stream tester deployed with metrics below:
      - Live feed from each Orchestrator (that opt in)
      - Segment Uptime
      - Cost  (USD)
      - Round Trip metrics (distance, up, down, delay)
   - Gather and host metrics with Prometheus (365 days) open ports for public viewing ( Aug 8, 2022 - use The Graph for indexing? )

**Title:** Ongoing costs
**Budget:** $2,400.00 USD* Estimated $200/mo
**Tasks rendered:**
- 1 year hosting VPS in New York region

**Phase 3:**
**Title:** Frontend
**Developer Hourly Wage:** $58/hr USD* Up for discussion 
**Estimated hours:** 40 hours
**Budget:** $2,320.00 USD
**Developers:** TBA
**Tasks rendered:**
- HTML, CSS, JAVASCRIPT front end
   - Lists all orchestrators that have opted in.
   - List the default selection algorithm as a stream as well. (to be figured out, include, how to include Orchestrators that have not opted in)
- Display metrics listed in Phase 2 of this application.

**Phase 4:** (nice to have)
**Title:** AWS stream
**Developer Hourly Wage:** $58/hr USD* Up for discussion 
**Estimated hours:** 80 hours
**Budget:** $4,640.00 USD
**Developers:** TBA
**Tasks rendered:**
- AWS transcoding service
   - Create program to stream and gather metrics
   - Host metrics with Prometheus (365 days) open ports for public viewing

**Phase 5:** To be completed
**Title:** Education
**Hourly Wage:** ???
**Estimated hours:** ???
**Budget:** ???
**Developers:** ???
**Tasks rendered:**
- Provide documentation, tutorials and content around understanding the tools in this program. (ie prometheus endpoints, metrics, livestream endpoints etc)

**Funds:**
The funds would pay for the following:
1. Phases
Each phase has an estimated time commitment developers will be compensated for their time.

2. In general technology costs (webhost, server costs, etc)
During Phase 2 a continuous cost for VPS hosting will incur and funds will be paid for 1 year in advance.

3. AWS stream service (if needed)
If we decide to complete Phase 4, a regular streaming cost to AWS transcoding service will incur, will be paid for 1 year in advance.

**Summary:**
Total grant application estimate excluding Phase 4: 
**$12,840 USD**

Total grant application estimate including Phase 4:
**$17,480.00 USD**

**Timeframe:**
Each phase executed in order.
Each phase will be reviewed upon completion of the previous phase and adjusted accordingly with updated time, costs and scope.
Assuming developers work part time (4 hours per day) each phase should take between 1-5 weeks.
Total time to complete the project is estimated at 12 weeks.

**Funding for Phase 1**
This initial grant application is for phase 1 funding paid directly to the developers listed.

Thank you for your consideration

**Comment by nelsorya** [comment]
Hey @Titan-Node, thanks for putting this together. This sounds like a great initiative that could be very helpful for the community. and great to see the community collaboration. Just to clarify, are there developers committed to phase 2 and 3 of the proposal? While we support the sentiment of the proposal, it is difficult to commit to funding the entire grant without developers committed to each milestone.

**Comment by Titan-Node** [comment]
Hi @nelsorya
We are still in the works for building out the phases. Unfortunately it will be hard to determine the hours needed and skills needed without phase 1 being finished. 
Are we able to fund phase 1 and re evaluate the phases as they come?
I've tried to break down the phases into deliverables that are valuable on their own so we can take one step at a time. 

If you require to fund an entire project as one approval then let me know. Maybe it is best if we break this down into multiple grants as we are trying to gain feedback from broadcasters as well as merge code with go-Livepeer which takes time. 

Let me know 
Thanks!

**Comment by nelsorya** [comment]
Hey @Titan-Node, ok that makes sense. We are happy to fund phase 1 and we can re-evaluate the phases as we go.

**Comment by ad-astra-video** [comment]
Hello, I would like to submit [livepeer-test-broadcaster.ad-astra.video](https://livepeer-test-broadcaster.ad-astra.video) for phases 2 and 3 of this grant.  This tool provides the ability to send test streams to specific orchestrators and tracks orchestrator performance over time using test streams every 5 minutes.  Data is parsed from broadcaster logs and summarized in the dashboard. Data is also available for download in JSON format (all stream data and data filtered to one orchestrator are available).

The community has found this tool valuable to provide information on running orchestrators and insight to network selection algorithm.   Some examples are adjusted T selection in pools (or removing Ts connected), moving orchestrator to a new server (or to combined O/T from split O/T), and using test stream requests to test new setups.

Some highlights:
- tool runs on mainnet and pays eth for test streams in two regions [USA-EAST](https://livepeer-test-broadcaster.ad-astra.video) and [EU](https://livepeer-test-broadcaster.eu.ad-astra.video)
- environment runs in docker containers using python, redis and nginx.
- go-livepeer broadcaster metrics are available at [/metrics](https://livepeer-test-broadcaster.ad-astra.video/metrics) endpoint. This can be consumed by anyone that wants to setup Prometheus to scrape the data.
- data provided includes transcoding performance (upload/transcoding time/download) and information on how orchestrators were selected

Enhancements on the roadmap for phase 4:
- accumulate data and provide a dashboard with performance over a longer time period (currently provides rolling 5 days of data)
- provide longer test streams for Os that have a price less than a specified amount (eg 100 wei/pixel)
- provide data in format Grafana can ingest (either Prometheus formatted metrics or via an endpoint using the RedisTimeSeries module)
- provide a dashboard of orchestrator up time and responsiveness
- provide notifications to orchestrators when performance is not optimal or the orchestrator is not responding (e.g. email, discord, telegram)
- Use catalyst as broadcaster component of the stack

**Comment by papabear99** [comment]
To add to what is posted above, not only has this tool already provided a lot of valuable data to all Os on the network.  It's the groundwork for the code Livepeer is implementing on their Bs to not only give us a view of Livepeer Inc's streams, but the data can be used to update the current selection process (initial discord conversation here https://discord.com/channels/423160867534929930/750796877762527262/1118623362382565546) to favor high reliability Os and skip non-performant  Os to improve network performance and reliability.

Additionally @ad-astra-video has been paying for the servers to host/run this and broadcast fees to send the tests.

**Comment by Titan-Node** [comment]
I second @ad-astra-video's contribution, well deserving of the grant for phase 2-3.

**Comment by hansy** [comment]
@ad-astra-video Thanks for the update! Love the idea for this kind of dashboard. So Phase 2 is the "Request Stream" tab, that gives a broadcaster information about a particular Orchestrator. Apologies for my ignorance but can you show me how to use this tool and what information I can get from it to help me compare to web2 services like AWS, Mux, etc?

Phase 3 would be on the "Selection Tests" tab, right? This is meant to generally compare performance of O's who have opted in? Do you mind also providing a little bit more information on the type of information being displayed here?

This is awesome work by the way.

**Comment by ad-astra-video** [comment]
@hansy 

You are correct the "Request Stream" tab is the more specific information for an orchestrator and is designed to let the user (usually an orchestrator) see the transcoded video results.  It also provides some information on the transcoding performance. for each segment.

For right now, my broadcaster is paying market rates for transcoding to get the test results.   This tool could be enhanced to provide longer stream tests for Os that set the transcoding price for this broadcaster to a much lower value (e.g. 25 wei/pixel) threshold to keep transcoding costs reasonable.  The lower wei/pixel would act as the "opt in" component.

> can you show me how to use this tool and what information I can get from it to help me compare to web2 services like AWS, Mux, etc?

I do not have this implemented yet.  I can add this to the Stream tab so the cost can be seen as the stream progresses.

The Request Stream tab is used by typing in eth address and clicking Request Test Stream. A live stream is then started and played back for the user to see the results.  The test results are then populated under the Test Results button after the stream ends.

> Phase 3 would be on the "Selection Tests" tab, right? This is meant to generally compare performance of O's who have opted in? Do you mind also providing a little bit more information on the type of information being displayed here?

Correct, this is a view of the livepeer network over a trailing 5 days for my broadcaster.  The data is summarized in the dashboard and the data is available for download with the Download Logs and Download Data buttons.  The summary information can be filtered to one orchestrator by putting the eth address/ens name in the search box.  

The summary data in the cards provides an overview of the network including information on the selection algorithm, the pools of orchestrators in selection and the overall transcoding performance.  The bar graph provides information on the selection algorithm as well by showing the orchestrators selected, how often and how the orchestrator was selected (stake weight vs random).  The bar graph also includes a couple transcoding performance components with the red bar showing the total swaps each orchestrator experienced due to poor performance and a grey bar showing transcoding performance.  The transcoding performance bar is off by default because is usually 0-1 which gets to be very small with a lot of selection tests.  Turning on only the Transcode Score bar provides an overview of the average transcode score for each orchestrator for the test streams transcoded by the orchestrator.

The bar graphs have contextual detail when the cursor hovers over the bars to provide some additional data.  For the green and orange bars in the selection tests graph the total selections for stake/random weight and the total pools the orchestrator was included in is provided.  The transcode score graph provides total segments for the bar and, when filtered to a specific orchestrator, the red bar hover detail provides the manifestIDs of the streams that had poor transcoding performance.
 
The detail data (json) can be downloaded in total and filtered to each orchestrator by typing in an eth address or ens name.  The detail data download includes much more information on each test stream.  For each stream the detail data includes orchestrators in the selection pool, the orchestrator selected for each segment, eth paid in total and the segment paid on and upload/transcode/download times for each segment in milliseconds.

**Comment by ad-astra-video** [comment]
@hansy is there anything waiting on me?

**Comment by hansy** [comment]
@ad-astra-video Here's the tx receipt for both Phases 2 + 3: https://arbiscan.io/tx/0x288ecfe1c7d1630fa22621d1deadab1ba2300b3201242be3e29a7e171e3d12d9.

Let me know if you have any questions!
